From 5e617924df0ddf2c67232b160ac64884f9c87855 Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:15:14 -0400
Subject: [net] sctp: Change sctp to implement csum_levels

Message-id: <1409652920-431-52-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90160
O-Subject: [PATCH 51/57] [RHEL7.1 net] sctp: Change sctp to implement csum_levels
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit 202863fe4c7a5b0b9a3d3a00d207691635b31930
Author: Tom Herbert <therbert@google.com>
Date:   Wed Aug 27 21:27:06 2014 -0700

    sctp: Change sctp to implement csum_levels

    CHECKSUM_UNNECESSARY may be applied to the SCTP CRC so we need to
    appropriate account for this by decrementing csum_level. This is
    done by calling __skb_dec_checksum_unnecessary.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sctp/input.c b/net/sctp/input.c
index dd6cc64..cdb16dc 100644
--- a/net/sctp/input.c
+++ b/net/sctp/input.c
@@ -140,9 +140,13 @@ int sctp_rcv(struct sk_buff *skb)
  __skb_pull(skb, skb_transport_offset(skb));
  if (skb->len < sizeof(struct sctphdr))
   goto discard_it;
- if (!sctp_checksum_disable && !skb_csum_unnecessary(skb) &&
-    sctp_rcv_checksum(net, skb) < 0)
+
+ skb->csum_valid = 0; /* Previous value not applicable */
+ if (skb_csum_unnecessary(skb))
+  __skb_decr_checksum_unnecessary(skb);
+ else if (!sctp_checksum_disable && sctp_rcv_checksum(net, skb) < 0)
   goto discard_it;
+ skb->csum_valid = 1;
 
  skb_pull(skb, sizeof(struct sctphdr));
 
-- 
1.7.1